Yank/yank.py

```diff
@@ -245,6 +251,237 @@ def get_region(self, region_name):
+            SORT_PRIORITY = ('ordered', 'inferred_by_region', None)
+
+            @classmethod
+            def select(cls, region_string):
+                try:
+                    # The self here is inherited from the outer scope
+                    region_output = list(self._get_region_set(region_string))
+                except (SyntaxError, ValueError) as e:
+                    # Make this a local variable
+                    region_error = e
+                    region_output = None
+                else:
+                    region_error = None
+                return region_output, region_error
+
+        class SelectDsl(Selector):
+            SORT_PRIORITY = ('ordered', None)
+
+            @classmethod
+            def select(cls, dsl_string):
+                try:
+                    mdtraj_output = (self.topology.select(dsl_string)).tolist()
+                except ValueError as e:
+                    # Make this a local variable
+                    mdtraj_error = e
+                    mdtraj_output = None
+                else:
+                    mdtraj_error = None
+                return mdtraj_output, mdtraj_error
+
+        class SelectSmarts(Selector):
+
+            @classmethod
+            def select(cls, smarts_string):
+
+                try:
+                    # Skeletal structure, not used yet
+                    raise NotImplementedError("This method has not been implemented yet")
+                except (NotImplementedError, ValueError) as e:
+                    smarts_error = e
+                    smarts_output = None
+                return smarts_output, smarts_error
+
+        class SelectIterable(Selector):
+            # Do not allow these to be sorted as they have been manually specified by integer
+
+            @classmethod
+            def select(cls, iterable):
+                try:
+                    iterable_output = iterable.tolist()
+                except AttributeError:
+                    iterable_output = list(iterable)
+                    iterable_error = None
+                except Exception as e:
+                    iterable_error = e
+                    iterable_output = None
+                else:
+                    iterable_error = None
+                return iterable_output, iterable_error
+
+        class SelectInt(SelectIterable):
+            # Verbatim copy of Iterable, but change name to help error processing
+            pass
+
+        # Dispatcher to parse the selection type and return the valid selection classes
+        @functools.singledispatch
+        def selector_picker(selection_input):
+            try:
+                # Catch both the numpy and built in integers
+                assert all([isinstance(i, np.integer) or isinstance(i, int) for i in selection_input])
+            except AssertionError:
+                raise AssertionError("Selection {} is not iterable of ints or any other readable type such as string! "
```

Comment:
I think you should use either a simple
```python
if not all(condition):
    raise ValueError()
``` 
or at most
```python
assert all(condition), 'Selection {} is not iterable ...'
```
`assert`s can be deactivated when calling the python interpreter.

Reply:
I am trying to keep cyclomatic complexity down, and this lets us much more easily extend the function later. I had actually thought about splitting the selection/sorting functionality into its own class, but that did not make sense since its only used in this function.

Reply:
I see. I think we should favor readability over lower cyclomatic complexity though. Again, this works perfectly so feel free to merge it, but I'd change the whole block with a more compact
```python
if isinstance(selection, int):
    registered_selectors = SelectInt,
elif isinstance(selection, str):
    registered_selectors = SelectRegion, SelectDsl, SelectSmarts
elif not all([isinstance(i, np.integer) or isinstance(i, int) for i in selection_input]):
    raise ValueError('...')
else:
    registered_selectors = SelectIterable,
```
I think it's just easier to understand what's going on, and we can sacrifice the extensibility since this is all hidden inside the function anyway.